fili-core/src/test/groovy/com/yahoo/bard/webservice/async/AsyncFunctionalSpec.groovy

```diff
@@ -0,0 +1,211 @@
+package com.yahoo.bard.webservice.async
+
+import com.yahoo.bard.webservice.application.JerseyTestBinder
+import com.yahoo.bard.webservice.models.druid.client.impl.TestDruidWebService
+import com.yahoo.bard.webservice.table.availability.AvailabilityTestingUtils
+import com.yahoo.bard.webservice.web.endpoints.BaseDataServletComponentSpec
+import com.yahoo.bard.webservice.web.endpoints.DataServlet
+import com.yahoo.bard.webservice.web.endpoints.JobsServlet
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module
+
+import org.joda.time.Interval
+
+import spock.lang.Specification
+import spock.lang.Timeout
+
+import javax.validation.constraints.NotNull
+import javax.ws.rs.core.Response
+
+
+/**
+ * Serves as a base class for testing asynchronous processes.
+ *
+ * This test allows us to execute a series of queries one after the other, and use the results of previous queries to
+ * build subsequent queries.
+ * <p>
+ * As such, this spec expects a map of closures rather than a single target. The test will do something similar to, but
+ * not quite a reduction over this map. First, it will execute the first closure in the map (note that we are using the
+ * fact that Groovy's maps have a well-defined ordering) to get a target. We will make a request to the test server
+ * using that target (plus query parameters). Then we will take the result of that query (a Response), verify it, and
+ * pass it into the next closure, which will build the next target to make a request against. We will get the result,
+ * verify it, and then pass it and the result from the first request to the third closure to get the third target to
+ * execute, and so on.
+ * <p>
+ * The keys of the map are arbitrary, but must be consistent across the maps of target builders, the map of
+ * result verifiers, and the map of query parameters within a single specification. Essentially, the keys
+ * serve as names for each request.
+ * <p>
+ * Note that this test is expected to be used for testing asynchronous interactions, _not_ that Bard constructs the
+ * correct Druid query for a given request. For tests of that nature, see the {@link BaseDataServletComponentSpec}.
+ *
+ * @param <T>  The type to use for keys into the maps containing test-specific resources
+ */
+@Timeout(30)
+abstract class AsyncFunctionalSpec extends Specification {
+
+    private static final ObjectMapper MAPPER = new ObjectMapper()
+            .registerModule(new Jdk8Module().configureAbsentsAsNulls(false))
+
+    JerseyTestBinder jtb
+
+    /**
+     * Returns a map of closures, each of which transforms a map of named Responses into request targets.
```

Comment:
Giving the expected method signature of these Closures (since the Closure type interface doesn't indicate it) would be good, I think. Based on this JavaDoc, it looks like it might be something like `Map<String, Response> -> requestTarget`, but it isn't clear.